pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go

```diff
@@ -2,34 +2,70 @@ package remote
 
 import (
 	"context"
+	"time"
 
 	"github.com/grafana/grafana/pkg/infra/log"
 	apimodels "github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
 	"github.com/grafana/grafana/pkg/services/ngalert/models"
 	"github.com/grafana/grafana/pkg/services/ngalert/notifier"
 )
 
+//go:generate mockery --name remoteAlertmanager --structname RemoteAlertmanagerMock --with-expecter --output mock --outpkg alertmanager_mock
+type remoteAlertmanager interface {
+	notifier.Alertmanager
+	CompareAndSendConfiguration(context.Context, *models.AlertConfiguration) error
+	CompareAndSendState(context.Context) error
+}
+
 type RemoteSecondaryForkedAlertmanager struct {
 	log log.Logger
 
 	internal notifier.Alertmanager
-	remote   notifier.Alertmanager
+	remote   remoteAlertmanager
+
+	lastSync     time.Time
+	syncInterval time.Duration
 }
 
-func NewRemoteSecondaryForkedAlertmanager(l log.Logger, internal, remote notifier.Alertmanager) *RemoteSecondaryForkedAlertmanager {
+func NewRemoteSecondaryForkedAlertmanager(l log.Logger, syncInterval time.Duration, internal notifier.Alertmanager, remote remoteAlertmanager) *RemoteSecondaryForkedAlertmanager {
 	return &RemoteSecondaryForkedAlertmanager{
-		log:      l,
-		internal: internal,
-		remote:   remote,
+		log:          l,
+		internal:     internal,
+		remote:       remote,
+		syncInterval: syncInterval,
 	}
 }
 
 // ApplyConfig will only log errors for the remote Alertmanager and ensure we delegate the call to the internal Alertmanager.
 // We don't care about errors in the remote Alertmanager in remote secondary mode.
 func (fam *RemoteSecondaryForkedAlertmanager) ApplyConfig(ctx context.Context, config *models.AlertConfiguration) error {
-	if err := fam.remote.ApplyConfig(ctx, config); err != nil {
-		fam.log.Error("Error applying config to the remote Alertmanager", "err", err)
+	var syncErr bool
```

Comment:
Good work here - as a note, I have a suggestion to simplify this code:

```git
diff --git a/pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go b/pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go
index 284de738b9..ffc27d98f5 100644
--- a/pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go
+++ b/pkg/services/ngalert/remote/remote_secondary_forked_alertmanager.go
@@ -8,6 +8,7 @@ import (
 	apimodels "github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
 	"github.com/grafana/grafana/pkg/services/ngalert/models"
 	"github.com/grafana/grafana/pkg/services/ngalert/notifier"
+	"golang.org/x/sync/errgroup"
 )
 
 //go:generate mockery --name remoteAlertmanager --structname RemoteAlertmanagerMock --with-expecter --output mock --outpkg alertmanager_mock
@@ -39,34 +40,40 @@ func NewRemoteSecondaryForkedAlertmanager(l log.Logger, syncInterval time.Durati
 // ApplyConfig will only log errors for the remote Alertmanager and ensure we delegate the call to the internal Alertmanager.
 // We don't care about errors in the remote Alertmanager in remote secondary mode.
 func (fam *RemoteSecondaryForkedAlertmanager) ApplyConfig(ctx context.Context, config *models.AlertConfiguration) error {
-	var syncErr bool
-	if !fam.remote.Ready() {
-		// If the Alertmanager has not been marked as "ready" yet, delegate the call to the remote Alertmanager.
-		// This will sync the Alertmanagers.
+	g, ctx := errgroup.WithContext(ctx)
+
+	// First, make sure we run the internal apply config - we only care about errors for this one.
+	g.Go(func() error {
+		return fam.internal.ApplyConfig(ctx, config)
+	})
+
+	// Then, make sure we try to figure out if we need to sync the external Alertmanager.
+	g.Go(func() error {
+		if fam.remote.Ready() && time.Since(fam.lastSync) >= fam.syncInterval {
+			// If the Alertmanager was marked as ready but the sync interval has elapsed, sync the Alertmanagers.
+			fam.log.Debug("Syncing configuration and state with the remote Alertmanager", "lastSync", fam.lastSync)
+			if err := fam.remote.CompareAndSendConfiguration(ctx, config); err != nil {
+				fam.log.Error("Unable to upload the configuration to the remote Alertmanager", "err", err)
+			}
+			if err := fam.remote.CompareAndSendState(ctx); err != nil {
+				fam.log.Error("Unable to upload the state to the remote Alertmanager", "err", err)
+			}
+			fam.log.Debug("Finished syncing configuration and state with the remote Alertmanager")
+
+			fam.lastSync = time.Now()
+			return nil
+		}
+
+		// If we made it this far it means that either remote is not ready or is not time to sync and, it is safe to call `ApplyConfig`.
 		if err := fam.remote.ApplyConfig(ctx, config); err != nil {
 			fam.log.Error("Error applying config to the remote Alertmanager", "err", err)
-			syncErr = true
 		}
-	} else if time.Since(fam.lastSync) >= fam.syncInterval {
-		// If the Alertmanager was marked as ready but the sync interval has elapsed, sync the Alertmanagers.
-		fam.log.Debug("Syncing configuration and state with the remote Alertmanager", "lastSync", fam.lastSync)
-		if err := fam.remote.CompareAndSendConfiguration(ctx, config); err != nil {
-			fam.log.Error("Unable to upload the configuration to the remote Alertmanager", "err", err)
-			syncErr = true
-		}
-		if err := fam.remote.CompareAndSendState(ctx); err != nil {
-			fam.log.Error("Unable to upload the state to the remote Alertmanager", "err", err)
-			syncErr = true
-		}
-		fam.log.Debug("Finished syncing configuration and state with the remote Alertmanager")
-	}
 
-	// If there were no errors syncing the configuration/state, update lastSync
-	if !syncErr {
 		fam.lastSync = time.Now()
-	}
+		return nil
+	})
 
-	return fam.internal.ApplyConfig(ctx, config)
+	return g.Wait()
 }
 
 // SaveAndApplyConfig is only called on the internal Alertmanager when running in remote secondary mode.

```

There's two things here -

1. I think we should execute this operations in parallel - for remote we make both a database call and an HTTP call which can result in significant delays for `ApplyConfig`. I think it's fine blocking given `ApplyConfig` is called within a goroutine.
2. The refactor of the conditionals which is entirely optional.